scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+        return u, s[:n+1], rc_f, info
+    if typecode == cnp.NPY_DOUBLE:
+        rc_d = rcond
+        info = reorth(m, n, <double*>qp, qs, qisF, <double*>up, us, <double*>sp, &rc_d)
+        return u, s[:n+1], rc_d, info
+    if typecode == cnp.NPY_CFLOAT:
+        rc_f = rcond
+        info = reorth(m, n, <float_complex*>qp, qs, qisF, <float_complex*>up, us, <float_complex*>sp, <float_complex*>&rc_f)
+        return u, s[:n+1], rc_f, info
+    if typecode == cnp.NPY_CDOUBLE:
+        rc_d = rcond
+        info = reorth(m, n, <double_complex*>qp, qs, qisF, <double_complex*>up, us, <double_complex*>sp, <double_complex*>&rc_d)
+        return u, s[:n+1], rc_d, info
+
+cdef int reorth(int m, int n, blas_t* q, int* qs, bint qisF, blas_t* u,
+                int* us, blas_t* s, blas_t* RCOND) nogil:
+    """Given a (m,n) matrix q with orthonormal columns and a (m,) vector u,
+       find vectors s, w and scalar p such that u = Qs + pw where w is of unit
+       length and orthogonal to the columns of q.
+
+       FIX comment on return values, and RCOND 
+
+       The method used for orthogonalizing u against q is described in [5]
+       listed in the file header.
+    """
+    cdef blas_t unorm, snorm, wnorm, wpnorm, sigma_max, sigma_min, rc
+    cdef char* T = 'T'
+    cdef char* N = 'N'
+    cdef char* C = 'C'
+    cdef int ss = 1
+    cdef blas_t inv_root2 = <blas_t>sqrt(2)
+
+    # normalize u
+    unorm = nrm2(m, u, us[0])
+    scal(m, 1/unorm, u, us[0])
+
+    # decompose u into q's columns.
+    if qisF:
+        if blas_t is float or blas_t is double:
+            gemv(T, m, n, 1, q, qs[1], u, us[0], 0, s, 1)
+        else:
+            gemv(C, m, n, 1, q, qs[1], u, us[0], 0, s, 1)
+    else:
+        if blas_t is float or blas_t is double:
+            gemv(N, n, m, 1, q, n, u, us[0], 0, s, 1)
+        else:
+            blas_t_conj(m, u, us)
+            gemv(N, n, m, 1, q, n, u, us[0], 0, s, 1)
+            blas_t_conj(m, u, us)
+            blas_t_conj(n, s, &ss)
+
+    # sigma_max is the largest singular value of q augmented with u/unorm
+    snorm = nrm2(n, s, 1)
+    sigma_max = blas_t_sqrt(1 + snorm)
+
+    # make u be the part of u that is not in span(q)
+    # i.e. u -= q.dot(s)
+    if qisF:
+        gemv(N, m, n, -1, q, qs[1], s, 1, 1, u, us[0])
+    else:
+        gemv(T, n, m, -1, q, n, s, 1, 1, u, us[0])
+    wnorm = nrm2(m, u, us[0])
+
+    # sigma_min is the smallest singular value of q qugmented with u/unorm
+    # the others are == 1, since q is orthonormal.
+    sigma_min = wnorm / sigma_max
+    rc = sigma_min / sigma_max
+
+    # check the conditioning of the problem.
+    if blas_t_less_than(rc, RCOND[0]):
+        RCOND[0] = rc
+        return 2
+    RCOND[0] = rc
+
+    if blas_t_less_than(inv_root2, wnorm):
+        scal(m, 1/wnorm, u, us[0])
+        scal(n, unorm, s, 1)
+        s[n] = unorm*wnorm
+        return 0
+
+    # if the above check failed, try one reorthogonalization
```

Comment:
I read this the wrong way around "check failed" in >50% of the time means "enter the if-block and do something to fix it". it's a minor thing, but could be clarified a bit. Or in the block above add a comment that the ``return 0`` is the main path, and the rest happens in case the check (which does ...) failed.